### PR TITLE
#831 Migrate from SQLiteOpenHelper to Room

### LINKED ISSUE #831: Migrate from SQLiteOpenHelper to Room
- [ ] Add Entities and Daos for `Account` and `User`
- [ ] Create methods to restore compatibility with current callers of SQLiteOpenHelper
- [ ] Migrate existing database
- [ ] If possible rename database to `NOTES.db` while migrating
- [ ] Restore old migration steps

*No goal*

- It is *not* a goal to move to `LiveData` in this step

### LINKED ISSUE #1023: Items in double when starred
Please use **[GitHub reactions](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 to show that you are affected by the same issue. Please don't comment if you have no relevant information to add!

**Describe the bug**
A note starred is duplicated in the list view

**To Reproduce**
1. Go to 'All the Notes'
2. Click on a star not activated
3. Scroll a bit
4. See error: the note is starred and not

**Expected behavior**
Only the starred should be here

**Screenshots**
x | x | x
--- | --- | ---
Screenshot_20210105-183501 | Screenshot_20210105-183517 | Screenshot_20210105-183529
Screenshot_20210105-183539 | Screenshot_20210105-183602

**Smartphone:**
 - Nextcloud Notes-Version (android app): 2.18.1
 - F-Droid or Play Store: F-Droid
 - Android-Version: 7
 - Device: Nexus5

**Server**
 - Nextcloud version: 19.0.4
 - Nextcloud Notes version (server app): 3.6.4

Thank you very much for your work guys and happys new year! :)

### PR DESCRIPTION
**Pull request by stefan-niedermann**
# 🏁 Goals

- Fix #831
- Fix #1023

# ✔️ Remaining tasks

- [x] Create `Entities`, `Daos`, etc and replace methods from `NotesDatabase` with methods from `NotesRoomDatabase`
- [x] Check where we can optimize performance again (like loading not the content of a note)
- [x] Make sync work again properly
- [x] Migration
- [x] Share into new note produces category with `null` which causes a crash
  - [x] Check where `NOT NULL` constraints and default values make sense
- [x] Check which database calls are performed at the main thread
  - `BaseNoteFragment#saveNote()` is still left but needs to stay synchronous for now (is documented in the source code)
- [x] Use [`Saved state module`](https://developer.android.com/topic/libraries/architecture/viewmodel-savedstate) in all ViewModels
- [x] Fix unit tests
- [x] `NoSuchElementException` for widgets
- [x] Update migration script
- [x] Migrate default Category sorting mechanism
- [x] Check "Sync error: No network connection" errors with disabled wi-fi and enabled "Sync only on wi-fi" setting
  - State is not optimal, but not different from `master`, so moving this task to a separate issue
- [x] Sharing from the list view does nothing
- [x] Add `IF EXISTS` to migration when dropping indizes
- [x] When creating a new note, the keyboard should be shown

# ☐ Follow-Up issues

- [ ] `EditActivity` should use an `EditViewModel` as abstraction and not directly access the database
- [ ] Check whether dialogs can (or should) share the corresponding ViewModels with the Activity
- [ ] Split `Repository` from the database class which coordinates access to the `SyncHelper` and the database for easier testing
- [ ] *Optional:* Use [`Retrofit`](https://square.github.io/retrofit/)
- [ ] *Optional:* Activate background synchronization by default
- [ ] *Optional:* Store category sorting method for `RECENT`, `FAVORITES` and `UNCATEGORIZED` per account
- [ ] *Optional:* [Edit title should not change a notes `modified` date](https://github.com/stefan-niedermann/nextcloud-notes/pull/967#pullrequestreview-633042274)

### CODE REVIEW

**Comment by korelstar** [comment]
I fixed some major issues, but this needs more testing.

I think there is still much wrong regarding the category handling: sometimes, the category title is used instead of the category id and vice versa. In order to make your life easier, **I strongly suggest not to use an artificial id for categories**. Instead, you could use the category string directly in the notes table. In the category table, the category string would be primary key (or unique key if that doesn't work with SQLite) (together with the account id).

**Comment by stefan-niedermann** [comment]
Awesome, thank you very, very much! :partying_face: 

I am aware that it makes a lot of trouble to use the artificial ID for the categories. We decided to go this way a few releases ago while implementing #814.

>  Instead, you could use the category string directly in the notes table. In the category table, the category string would be primary key (or unique key if that doesn't work with SQLite) (together with the account id). This will also improve performance, because you won't need to read from the category table if you don't need the additional attributes.

This definitely sounds like a good idea, i will try to implement it this way.

**Comment by stefan-niedermann** [comment]
# Call for testing

https://help.nextcloud.com/t/help-us-testing-the-refactored-notes-android-app/112828

@Isaac-Graham @beatbrot @dan0xii @jancborchardt @korelstar @David-Development

**Comment by beatbrot** [comment]
> 
> # Call for testing
> 
> https://help.nextcloud.com/t/help-us-testing-the-refactored-notes-android-app/112828
> 
> @Isaac-Graham @beatbrot @dan0xii @jancborchardt @korelstar @David-Development @IF-ACT @WingsUpete @Hui-Ouyang16 @fxedel @TacoTheDank @schaarsc @thgoebel @tobiasKaminsky

Unfortunately, I currently don't really have the time to review or test this PR. Sorry 😞

**Comment by Isaac-Graham** [comment]
> # Call for testing
> https://help.nextcloud.com/t/help-us-testing-the-refactored-notes-android-app/112828
> 
> @Isaac-Graham @beatbrot @dan0xii @jancborchardt @korelstar @David-Development @IF-ACT @WingsUpete @Hui-Ouyang16 @fxedel @TacoTheDank @schaarsc @thgoebel @tobiasKaminsky

I am so sorry but I am busy with my graduate thesis during these months, and I have no enough time to test these PRs.

Let me ask my classmates whether they are available.

**Comment by korelstar** [comment]
Got the following exception:
```
    Process: it.niedermann.owncloud.notes.dev, PID: 5093
    android.database.sqlite.SQLiteException: no such index: ACCOUNTS_URL_idx (code 1): , while compiling: DROP INDEX ACCOUNTS_URL_idx
        at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
        at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:889)
        at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:500)
        at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
        at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:58)
        at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java:31)
        at android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1677)
        at android.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1608)
        at androidx.sqlite.db.framework.FrameworkSQLiteDatabase.execSQL(FrameworkSQLiteDatabase.java:242)
        at it.niedermann.owncloud.notes.persistence.migration.Migration_20_21.dropOldIndices(Migration_20_21.java:43)
        at it.niedermann.owncloud.notes.persistence.migration.Migration_20_21.migrate(Migration_20_21.java:28)
        at androidx.room.RoomOpenHelper.onUpgrade(RoomOpenHelper.java:99)
        at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.onUpgrade(FrameworkSQLiteOpenHelper.java:177)
        at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:299)
        at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:194)
        at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase(FrameworkSQLiteOpenHelper.java:145)
        at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase(FrameworkSQLiteOpenHelper.java:106)
        at androidx.room.RoomDatabase.inTransaction(RoomDatabase.java:476)
        at androidx.room.RoomDatabase.assertNotSuspendingTransaction(RoomDatabase.java:281)
        at it.niedermann.owncloud.notes.persistence.dao.AccountDao_Impl.getAccountByName(AccountDao_Impl.java:582)
        at it.niedermann.owncloud.notes.persistence.NotesServerSyncHelper$1.lambda$onReceive$0$NotesServerSyncHelper$1(NotesServerSyncHelper.java:82)
        at it.niedermann.owncloud.notes.persistence.-$$Lambda$NotesServerSyncHelper$1$qBdFOVWkLK9cT0-oMt7eMhHEL24.run(Unknown Source:4)
        at java.lang.Thread.run(Thread.java:764)
```

I suggest to add "IF EXISTS" to all "DROP INDEX" statements. Everywhere.

**Comment by korelstar** [comment]
Hmm. After fixing that, I got the following error:

```
    java.lang.RuntimeException: Exception while computing database live data.
        at androidx.room.RoomTrackingLiveData$1.run(RoomTrackingLiveData.java:92)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: android.database.sqlite.SQLiteException: table `Account` already exists (code 1): , while compiling: CREATE TABLE `Account` (`id` INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, `url` TEXT NOT NULL DEFAULT '', `userName` TEXT NOT NULL DEFAULT '', `accountName` TEXT NOT NULL DEFAULT '', `eTag` TEXT, `modified` INTEGER, `apiVersion` TEXT, `color` INTEGER NOT NULL DEFAULT -16743735, `textColor` INTEGER NOT NULL DEFAULT -16777216, `capabilitiesETag` TEXT)
        at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
        at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:889)
        at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:500)
        at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
        at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:58)
        at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java:31)
        at android.database.sqlite.SQLiteDatabase.executeSql(SQLiteDatabase.java:1677)
        at android.database.sqlite.SQLiteDatabase.execSQL(SQLiteDatabase.java:1608)
        at androidx.sqlite.db.framework.FrameworkSQLiteDatabase.execSQL(FrameworkSQLiteDatabase.java:242)
        at it.niedermann.owncloud.notes.persistence.migration.Migration_20_21.createNewTables(Migration_20_21.java:57)
        at it.niedermann.owncloud.notes.persistence.migration.Migration_20_21.migrate(Migration_20_21.java:30)
```

Maybe it's because I had an old dev version installed? I will clear the app data, but I think the migration process should be tested more intensive.

**Comment by stefan-niedermann** [comment]
> I suggest to add "IF EXISTS" to all "DROP INDEX" statements. 

Good catch, i will add this.

I will also check the migration, though it should (TM) work fron the latest master and the latest released version 🙂

**Comment by korelstar** [comment]
last-modified / pruneBefore does not work. I tried to fix it, but it's hard to do this, because you use three different types of modified attributes and - that's the problem - they are mixed. Sometimes you use seconds, sometimes miliseconds, sometimes `Calendar`. I suggest to consequently use seconds (`long`) and never miliseconds and never `Calendar`. That would be much easier.

**Comment by stefan-niedermann** [comment]
I'd rather try to use `Calendar` in the database for all time related data and calculate to seconds in the `NotesAPI`. The detail level (= seconds) is API related. Semantically it's a date type.

In the future i plan to migrate all date related stuff to `Instant` as we did in the Deck Android app.

Will have a look at thus as well :)

File: app/src/main/java/it/niedermann/owncloud/notes/shared/model/CloudNote.java [hunk 1]
```diff
@@ -1,104 +0,0 @@
...
-    public void setCategory(String category) {
-        this.category = category == null ? "" : category;
-    }
-
-    @NonNull
-    @Override
-    public String toString() {
-        return "R" + getRemoteId() + " " + (isFavorite() ? " (*) " : "     ") + getCategory() + " / " + getTitle() + " (" + getModified(DATE_FORMAT) + " / " + getEtag() + ")";
-    }
-}
```
Comment (korelstar) [inline app/src/main/java/it/niedermann/owncloud/notes/shared/model/CloudNote.java:104]: This file is empty, but still exists. Maybe it should be deleted?

**Review by korelstar — CHANGES_REQUESTED** [review]
- [x] Last-modified / pruneBefore does not work (see comment above).
- [ ] Edit title should not change note's `modified` time.
- [x] There is no icon if a note's local changes are not synchronized to the server (e.g. when offline) anymore. Is this intended? I think it is an important indicator.
- [ ] when offline: `android.accounts.NetworkErrorException` (`NetworkInfo is null`) at it.niedermann.owncloud.notes.persistence.NotesServerSyncHelper.updateNetworkStatus(NotesServerSyncHelper.java:306), at it.niedermann.owncloud.notes.persistence.NotesServerSyncHelper$1.onReceive(NotesServerSyncHelper.java:78)

**Comment by stefan-niedermann** [comment]
@korelstar i think i found the cause of `pruneBefore` not working anymore:

https://github.com/stefan-niedermann/nextcloud-notes/pull/967/commits/20b4163ea014968f5daa95117b4217f2e286580d#diff-91dc585ffcf80718fa3006feb6f257fa80cf3c98709bad762330dd6bdaa5e466R197

I divided here by `1_000` which was wrong - since i don't get seconds or milliseconds here from the API but a `GMT` formatted string, the division was wrong.

`pruneBefore` should work now :slightly_smiling_face: 

About **changing the title**: Why not? The note was modified and the title will be synchronized back to the server, no? (At least for API 1.0+)

Regarding the **`NetworkErrorException`** - has it popped up as a dialog? Did the app crash? I can't reproduce a stacktrace. I get the "Synchronization failed: No network connection" snackbar, but as far as i can see this issue is also in the current master (as i stated in the very first comment of this PR) - therefore i'd like to not mess around with another refactoring but fix that once we have merged the Room migration.

**Comment by korelstar** [comment]
> `pruneBefore` should work now slightly_smiling_face

confirmed :smiley: 

-----

> About **changing the title**: Why not? The note was modified and the title will be synchronized back to the server, no? (At least for API 1.0+)

Indeed, this can be seen differently. I argue that the title is a meta information, just like the category. In both cases, the modified time should not change. The modified time should change only if the data (i.e. note's content) has changed, but not if only meta data has changed. This is how the notes server app implements this.

-----

> Regarding the **`NetworkErrorException`** - has it popped up as a dialog? Did the app crash? I can't reproduce a stacktrace. I get the "Synchronization failed: No network connection" snackbar, but as far as i can see this issue is also in the current master (as i stated in the very first comment of this PR) - therefore i'd like to not mess around with another refactoring but fix that once we have merged the Room migration.

No crash, no popup. Just saw it in Logcat. Nevertheless, it looks like the offline mode isn't correctly detected. On manual sync (pull to sync), I see the snackbar `Sychronization failed: No network connection`. However, the active sync animation starts, but is never hidden. Only if I activate the internet connection, the animations hides (after synchronization has finished). But this is not related with "Sync only on Wi-Fi" as stated in your first posting. Logcat shows this exception:

```
    com.nextcloud.android.sso.exceptions.UnknownErrorException: Unable to resolve host "XXX": No address associated with hostname
        at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:172)
        at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:171)
        at it.niedermann.nextcloud.sso.glide.SingleSignOnStreamFetcher.loadData(SingleSignOnStreamFetcher.kt:68)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.loadData(MultiModelLoader.java:100)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.startNextOrFail(MultiModelLoader.java:164)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.onLoadFailed(MultiModelLoader.java:154)
        at it.niedermann.nextcloud.sso.glide.SingleSignOnStreamFetcher.loadData(SingleSignOnStreamFetcher.kt:83)
        at com.bumptech.glide.load.model.MultiModelLoader$MultiFetcher.loadData(MultiModelLoader.java:100)
        at com.bumptech.glide.load.engine.SourceGenerator.startNextLoad(SourceGenerator.java:70)
        at com.bumptech.glide.load.engine.SourceGenerator.startNext(SourceGenerator.java:63)
        at com.bumptech.glide.load.engine.DecodeJob.runGenerators(DecodeJob.java:310)
        at com.bumptech.glide.load.engine.DecodeJob.runWrapped(DecodeJob.java:279)
        at com.bumptech.glide.load.engine.DecodeJob.run(DecodeJob.java:234)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
        at com.bumptech.glide.load.engine.executor.GlideExecutor$DefaultThreadFactory$1.run(GlideExecutor.java:393)
```

**Comment by stefan-niedermann** [comment]
> Indeed, this can be seen differently. I argue that the title is a meta information, just like the category. In both cases, the modified time should not change. The modified time should change only if the data (i.e. note's content) has changed, but not if only meta data has changed. This is how the notes server app implements this.

Marked this point as checked because it is also the current behavior in the `master` branch → Not blocking this PR. I added it to the [list of possible "Follow-Up issues"](https://github.com/stefan-niedermann/nextcloud-notes/pull/967#issue-497771696). Though in my understanding, the title was a `meta` information at API `0.2`, as it was generated, just like the `excerpt`. Since API `1.0` it is something that a user explicitly changes by intention, so we should talk about this topic again to get a common understanding.

Will have a look at the `OfflineException` now :slightly_smiling_face:

**Comment by stefan-niedermann** [comment]
The `NetworkErrorException` is catched properly and logged on purpose - no issue in this method.

Root cause is that the capabilities will be queried before the notes are synchronized and after this fails (when offline), the refresh spinner is not resetted. I will have a further look.

**Comment by stefan-niedermann** [comment]
@korelstar I fixed the remaining issues regarding the synchronization mechanism.

Be aware, that a dirty **update from an older state of this branch will fail**, because i had to slightly adjust the database to fix an issue with the Notes List Widget.

The migration from the latest `master` is tested and still works.

**Comment by korelstar** [comment]
There is still a problem with `pruneBefore`. If I filter logcat for `pruneBefore`, I get the following:

```
2021-04-21 07:26:02.511 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["76e12a0b73f5c80097d9bca4f998b31e"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:26:07.856 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["76e12a0b73f5c80097d9bca4f998b31e"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:26:11.480 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["76e12a0b73f5c80097d9bca4f998b31e"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:27:15.806 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["76e12a0b73f5c80097d9bca4f998b31e"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:27:18.631 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["f1f4be51638c0b1d114e68d330c85921"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:27:21.203 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["f1f4be51638c0b1d114e68d330c85921"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:27:29.367 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["f1f4be51638c0b1d114e68d330c85921"]}, parameter={pruneBefore=1618982836}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:27:34.447 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["c5cd9e408c82d814206ed99b1e109be5"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:31:22.281 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["c5cd9e408c82d814206ed99b1e109be5"]}, parameter={pruneBefore=1618982851}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
2021-04-21 07:31:26.788 5628-5750/it.niedermann.owncloud.notes.dev D/NotesClient: NextcloudRequest: NextcloudRequest(method=GET, header={If-None-Match=["ebd1ef9504d29e8871e64a71152992d1"]}, parameter={pruneBefore=1618982751}, requestBody=null, url=/index.php/apps/notes/api/v1/notes, token=null, packageName=null, accountName=null, bodyAsStream=null, followRedirects=false)
```

- In the beginning, I did some refreshes and `pruneBefore` doesn't change (`pruneBefore=1618982751`) => correct
- Then I changed a note, and `pruneBefore` was updated (`pruneBefore=1618982836`) => correct
- Then I did a manual refresh and `pruneBefore` has the old value from the beginning (`pruneBefore=1618982751`) => not correct!
- I repeated and the same problem re-occurs (last two lines)

**Comment by stefan-niedermann** [comment]
I think i know where the issue might come from...

We had a similar issue with an outdated `ETag` because we used an `Account` object which wasn't updated. We worked around this issue by reading the `ETag` from the database again:

https://github.com/stefan-niedermann/nextcloud-notes/blob/831-room/app/src/main/java/it/niedermann/owncloud/notes/persistence/NotesServerSyncTask.java#L167

I will try to also use the `modified` date from the newly read `Account` object, it is likely outdated for the same reason.

Will notify you once i changed it. Thanks for testinf again!

**Review by korelstar — APPROVED** [review]

**Pull request merged**